lib/charms/sdcore_nrf/v0/fiveg_nrf.py

```diff
@@ -0,0 +1,281 @@
+from ops.framework import EventBase, EventSource, Handle, Object  # noqa: E402
+from pydantic import AnyHttpUrl, BaseModel, Field, ValidationError  # noqa: E402
+
+logger = logging.getLogger(__name__)
+
+"""Schemas definition for the provider and requirer sides of the `fiveg_nrf` interface.
+It exposes two interfaces.schema_base.DataBagSchema subclasses called:
+- ProviderSchema
+- RequirerSchema
+Examples:
+    ProviderSchema:
+        unit: <empty>
+        app: {"url": "https://nrf-example.com:1234"}
+    RequirerSchema:
+        unit: <empty>
+        app:  <empty>
+"""
+
+
+class MyProviderAppData(BaseModel):
+    """Provider app data for fiveg_nrf."""
+
+    url: AnyHttpUrl = Field(
+        description="Url to reach the NRF.", examples=["https://nrf-example.com:1234"]
+    )
+
+
+class ProviderSchema(DataBagSchema):
+    """Provider schema for fiveg_nrf."""
+
+    app: MyProviderAppData
+
+
+class RequirerSchema(DataBagSchema):
+    """Requirer schema for fiveg_nrf."""
+
+
+class NRFAvailableEvent(EventBase):
+    """Charm event emitted when a NRF is available. It carries the NRF url."""
+
+    def __init__(self, handle: Handle, url: str):
+        """Init."""
+        super().__init__(handle)
+        self.url = url
+
+    def snapshot(self) -> dict:
+        """Returns snapshot."""
+        return {"url": self.url}
+
+    def restore(self, snapshot: dict) -> None:
+        """Restores snapshot."""
+        self.url = snapshot["url"]
+
+
+class NRFRequirerCharmEvents(CharmEvents):
+    """List of events that the NRF requirer charm can leverage."""
+
+    nrf_available = EventSource(NRFAvailableEvent)
+
+
+class NRFRequires(Object):
+    """Class to be instantiated by the NRF requirer charm."""
+
+    on = NRFRequirerCharmEvents()
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Init."""
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    @staticmethod
+    def _relation_data_is_valid(relation_data: dict) -> bool:
+        """Returns whether URL is valid.
+
+        Args:
+            str: URL to be validated.
+
+        Returns:
+            bool: True if URL is valid, False otherwise.
+        """
+        try:
+            ProviderSchema(app=relation_data)
+            return True
+        except ValidationError:
+            return False
+
+    def _on_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handler triggered on relation changed event.
+
+        Args:
+            event: Juju event (RelationChangedEvent)
+
+        Returns:
+            None
+        """
+        relation = self.model.get_relation(self.relation_name)
+        if not relation:
+            logger.warning(f"No relation: {self.relation_name}")
+            return
+        relation = event.relation
```

Comment:
why are we setting the value again here?

Reply:
Not anymore, Fixed.